PluginFactory should default to `dotnet` in the path to start a plugin if DOTNET_HOST_PATH is not set

olivier-spinelli opened the pull request:
### Bug
Fixes: NuGet/Home#7438
Regression: No

Last working version:
How are we preventing it in future:

### Fix
Simple fallback on "dotnet.exe" or "dotnet" when DOTNET_HOST_PATH environment variable is not defined.

Testing/Validation
Tests Added: No
Reason for not adding tests: Relying on the existing tests.

olivier-spinelli committed:
Fixed https://github.com/NuGet/Home/issues/7438

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs:
```diff
@@ -165,7 +165,8 @@ public void Dispose()
             {
-                FileName = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH"),
+                FileName = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH")
+                            ?? (RuntimeEnvironmentHelper.IsWindows ? "dotnet.exe" : "dotnet"),
```
Can format this a bit better, put it on a new line to improve readability, see Rule #15 https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md

olivier-spinelli replied:
Sorry! Would the following be ok? (Applied "Braces may be omitted only if the body of every block associated with an if/else if/.../else compound statement is placed on a single line.")
```csharp
                FileName = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH")
                                    ?? (RuntimeEnvironmentHelper.IsWindows 
                                           ? "dotnet.exe" 
                                           : "dotnet")
```

nkolev92 replied:
Yeah that's fine, I'd keep the operators on the previous line, but that's just my personal preference :D

nkolev92 replied:
While we are at, can you please use the PR template? 

We have automation that depends on it.

olivier-spinelli replied:
You're the boss ;)

```csharp
FileName = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH") ??
                (RuntimeEnvironmentHelper.IsWindows ? 
                     "dotnet.exe" :   
                     "dotnet")
```
Better?

olivier-spinelli replied:
Yes of course for the PR template. Sorry again.

nkolev92 replied:
Looks great, thanks!
I will kick off the CI. 

Please update the PR template.

olivier-spinelli committed:
Fixed coding style.

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs:
```diff
@@ -165,7 +165,10 @@ public void Dispose()
             var startInfo = new ProcessStartInfo
             {
-                FileName = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH"),
+                FileName = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH") ??
```
Compile error:   Plugins\PluginFactory.cs(169,29): error CS0103: The name 'RuntimeEnvironmentHelper' does not exist in the current context [E:\A\_work\290\s\src\NuGet.Core\NuGet.Protocol\NuGet.Protocol.csproj]

olivier-spinelli replied:
My bad... I didn't want to build the whole stuff (configure.ps1/build.ps1...). 

I used the full name  `NuGet.Common.RuntimeEnvironmentHelper.IsWindows` to avoid a `using NuGet.Common` that would be useless when `IS_DESKTOP` (and would have need a ugly `#if !IS_DESKTOP` conditional). 

I eventually ran the tests locally. The tests fail but somewhere else...

![image](https://user-images.githubusercontent.com/1583378/78897118-f9428200-7a71-11ea-8e26-561b06bdad6e.png)

I'm sorry I wasted your time.

nkolev92 replied:
No worries. I'll rerun again, I'm not expecting those tests to be in any way related to this change.

olivier-spinelli committed:
Fixed missing namespace.

Using the full name  `NuGet.Common.RuntimeEnvironmentHelper.IsWindows`:
this avoids a `using NuGet.Common` that would be useless when
`IS_DESKTOP` (and would have need a ugly `#if !IS_DESKTOP` conditional).
